src/reconnector.ts

```diff
@@ -163,52 +149,53 @@ export class Reconnector {
     this._backoffActive = true;
   }
 
-  stop (error?: any) {
+  stop (error?: Error | string) {
     this.cleanupReconnect();
     this._stopHardReconnect(error);
   }
 
-  private _stopHardReconnect (error?: any) {
+  private _stopHardReconnect (error?: Error | string) {
     if (!this._hardReconnectRetryInfo) return;
 
+    // SANITY: this is here for cases where `_stopHardReconnect()`
+    //  was called before the promise from `retryPromise()` finishes
     if (error) {
-      this._hardReconnectRetryInfo.reject(error);
+      /* ensure we are always working with an Error object */
+      if (typeof error === 'string') {
+        error = new Error(error);
+      }
+
+      this._hardReconnectRetryInfo.cancel(error);
     } else {
-      this._hardReconnectRetryInfo.resolve();
+      this._hardReconnectRetryInfo.complete();
     }
 
-    clearInterval(this._hardReconnectRetryInfo.interval);
     this._hardReconnectRetryInfo = null;
   }
 
   private async _attemptHardReconnect () {
     /* if we aren't online, don't retry the new channel */
     if (!navigator.onLine) {
-      return this.client.logger.debug('Browser is offline. Not attempting to reconnect with new channel.');
+      throw new Error(OFFLINE_ERROR);
     }
 
-    try {
-      this.client.logger.info('Attempting to reconnect with new channel.');
-
-      await this.client.connect();
+    this.client.logger.info('Attempting to reconnect with new channel.');
 
-      this._stopHardReconnect();
-    } catch (error) {
-      /* this error comes from superagent for requests that timeout for network offline */
-      if (error.message.startsWith('Request has been terminated')) {
-        return this.client.logger.debug('request offline. attempting reconnect again', error);
-      }
-      /* error client throws timeouts */
-      else if (error.message.startsWith('Timeout: ')) {
-        return this.client.logger.debug(`Streaming-client timed out. attempting reconnect again: "${error.message}"`);
-      }
-      /* superagent retriable error */
-      else if (error && this._retryStatusCodes.has(error.status)) {
-        return this.client.logger.debug('Received HTTP status code eligible for retry. attempting reconnect again', error);
-      }
+    await this.client.connect();
+  }
 
-      /* if it is an error we can't retry, reject with it */
-      this._stopHardReconnect(error);
+  private _shouldRetryError (error: Error): boolean {
```

Comment:
If this returns `true`, it means we need to keep trying the hard reconnect. Main use cases: 
1. network goes offline
2. 429s